agate/agate/authorisation.py

```diff
@@ -0,0 +1,78 @@
+import requests
+from .caching import TokenCache
+from core.settings import ONYX_DOMAIN
+from datetime import timedelta
+from django.utils import timezone
+import json
+
+
+def check_project_authorized(auth, project):
+    """
+    Boolean telling if a provided authorization token is valid to view a project
+
+    The onyx API is queried to determine which projects the token is permitted to see.
+    If the onyx response is not a 200, the token is invalid.
+    Otherwise the list of projects is compared against the requested project
+    """
+    projects = _get_item(auth).projects_output
+    for a in json.loads(projects)["data"]:
+        if a["project"] == project:
+            return True
+    return False
+
+
+def find_site(auth):
+    """
+    String telling which site a provided authorization token originates from
+
+    The onyx API is queried to determine the profile of the token.
+    If the onyx response is not a 200, the token is invalid, and the empty sting is returned.
+    Otherwise site is returned
+    """
+    return _get_item(auth).site_output
+
+
+def check_authorized(auth, site, project):
+    """
+    Boolean telling whether a provided authorization token BOTH
+    + Originates from the site
+    + Is authorized to view the project
+    """
+    return (find_site(auth) == site) and (check_project_authorized(auth, project))
+
+
+def _get_item(auth):
+
+    time_one_hour_ago = timezone.now() - timedelta(hours=1)
+    try:
+        item = TokenCache.objects.get(token_hash=hash(auth))
```

Comment:
I reckon use a cryptographically secure hash function from `hashlib`, e.g. `sha256`:

```
import hashlib

hasher = hashlib.sha256()
hasher.update(auth.encode("utf-8"))
hash = hasher.hexdigest()
```

Reply:
I know this is resolved but just for some context, the built-in `hash` function is the one that's used for hashtable based objects (`dict`s and `set`s, I think). @tombch's right that the builtin [`hashlib` library](https://docs.python.org/3/library/hashlib.html).

Note the warning that MD5 and SHA1 are regarded as insecure.